Refactored assertAutodiff test to have better error message

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48627 Refactored OpInfo testing to support custom SampleInputs, added addmm to op_db to test
* #47696 Support for OpInfo-based testing for operators in JIT
* **#48567 Refactored assertAutodiff test to have better error message**
* #47695 Refactor existing JIT testing utils to enable new OpInfo test suite to reuse existing logic

Refactored assertAutodiff to tell exactly which nodes are causing the test to fail with suggestions on how to fix the failure

Differential Revision: [D25212435](https://our.internmc.facebook.com/intern/diff/D25212435)

Commit:
Refactored assertAutodiff test to have better error message

[ghstack-poisoned]

Commit:
Update on "Refactored assertAutodiff test to have better error message"




[ghstack-poisoned]

Commit:
Update on "Refactored assertAutodiff test to have better error message"


Differential Revision: [D25212435](https://our.internmc.facebook.com/intern/diff/D25212435)

[ghstack-poisoned]

Commit:
Update on "Refactored assertAutodiff test to have better error message"


Differential Revision: [D25212435](https://our.internmc.facebook.com/intern/diff/D25212435)

[ghstack-poisoned]

Commit:
Update on "Refactored assertAutodiff test to have better error message"


Differential Revision: [D25212435](https://our.internmc.facebook.com/intern/diff/D25212435)

[ghstack-poisoned]

Commit:
Update on "Refactored assertAutodiff test to have better error message"


Differential Revision: [D25212435](https://our.internmc.facebook.com/intern/diff/D25212435)

[ghstack-poisoned]

Lilyjjo commented:
<img width="1484" alt="Screen Shot 2020-11-30 at 3 59 59 PM" src="https://user-images.githubusercontent.com/35852084/100684458-3c1dd680-332f-11eb-8425-89af7b3f0afc.png">

Image of error message in action

Commit:
Update on "Refactored assertAutodiff test to have better error message"


Refactored assertAutodiff to tell exactly which nodes are causing the test to fail with suggestions on how to fix the failure

Differential Revision: [D25212435](https://our.internmc.facebook.com/intern/diff/D25212435)

[ghstack-poisoned]

eellison commented on torch/testing/_internal/common_jit.py:
```diff
@@ -143,18 +143,78 @@ def getExportImportCopy(self, m, also_test_file=True, map_location=None):
+                "but were not found in specified fusible/nonfusible " \
+                "DifferentiableGraph groups. \nSpecifically:"
+            for node in nodes_not_in_diff_graph:
+                err_msg += f"\n  {node} was not in one of the DifferentiableGraphs " \
```
it might be nice to join up this error message for all nodes that are not in when expected to be. 
`{aten::add, aten::sum} were not in one of the DifferentiableGraphs` and the same for bucketing all the `non_fusible_nodes_being_fused` `fusion_nodes_not_found` groups

eellison reviewed (approved):
looks great!!

mruberry commented on torch/testing/_internal/common_jit.py:
```diff
@@ -143,18 +143,78 @@ def getExportImportCopy(self, m, also_test_file=True, map_location=None):
+                err_msg += f"\n  {node} was not in one of the DifferentiableGraphs " \
+                    "when it was expected to be. Did you intend for this node to be " \
+                    "autodiffed? If not, remove it from the list of nonfusible nodes."
+                if node in non_fusible_nodes_being_fused:
```
If `should_autodiff_node` is True, a `foo` node is in `nodes_not_in_diff_graph` and that node is also in `non_fusible_nodes_being_fused`, then I think the printed error message will be:

> Failure in testing nodes' autodifferentiation, one or more nodes were expected to be autodiffed but were not found in specified fusible/nonfusible DifferentiableGraph groups. 
> 
> Specifically: 
> `foo` was not in one of the DifferentiableGraphs when it was expected to be. Did you intend for this node to be autodiffed? If not, remove it from the list of nonfusible nodes. Additionally, This [sic] node was found in a FusionGroup in a DifferentiableGraph. If that was intended, reclassify this node as a fusible node. If not, your autodifferentiation logic might be wrong.

To someone not familiar with the JIT, it seems like the above is saying that `foo` is both in and not in a differentiable graph. Since it's in a fusion group in a differentiable graph but the paragraph for `foo` starts by saying it's not in a differentiable graph. 

I think the cases for `should_autodiff_node=True` are:

- [Case 1] the node is intended to appear in a differentiable graph but doesn't
- [Case 2] the node is intended to appear in a differentiable graph outside of a fusion group, but it's in a fusion group
- [Case 3] the node is intended to appear in a differentiable graph inside a fusion group, but it's not inside a fusion group

Is that correct? If so, I would suggest identifying which case the op falls in and then crafting a simplified error message for each case. For example,

[Case 1] "Node `foo` is marked as being autodifferentiable, but doesn't appear in any differentiable graph. If this is expected, ..."
[Case 2] "Node `foo` is marked as being nonfusible, but appears in a fusion group. If `foo` is expected to be fused then remove it from the list of nonfusible nodes." 
[Case 3] "Node `foo` is marked as being fusible, but appears outside a fusion group. If `foo` is expected to be unfused then add it to the list of nonfusible nodes."

Does that makes sense or am I misunderstanding these cases?

Unfortunately this proposal is much easier to implement per-op, which makes implementing a suggestion like @ellison's above a little trickier. The nodes could be added to three lists, one for each category, before the error messages are constructed, however.

mruberry commented on torch/testing/_internal/common_jit.py:
```diff
@@ -143,18 +143,78 @@ def getExportImportCopy(self, m, also_test_file=True, map_location=None):
+                    "when it was expected to be. Did you intend for this node to be " \
+                    "autodiffed? If not, remove it from the list of nonfusible nodes."
+                if node in non_fusible_nodes_being_fused:
+                    err_msg += "Additionally, This node was found in a FusionGroup " \
```
"This" -> "this"

mruberry commented on torch/testing/_internal/common_jit.py:
```diff
@@ -143,18 +143,78 @@ def getExportImportCopy(self, m, also_test_file=True, map_location=None):
+                err_msg += f"\n  {node} was not in one of the DifferentiableGraphs' " \
+                    "fusion groups when it was expected to be. " \
+                    "Did you intend for this node to be fused? If not, you should " \
+                    "move this node into the test's non-fusible nodes."
```
"non-fusible" here is inconsistent with "nonfusible" above.

mruberry commented on torch/testing/_internal/common_jit.py:
```diff
@@ -143,18 +143,78 @@ def getExportImportCopy(self, m, also_test_file=True, map_location=None):
+                "If so, change this test to expect autodifferention. " \
+                "\nSpecifically:"
+            for node in fusion_nodes_found:
+                err_msg += f"\n  {node} was not expected to in one of the " \
```
"to in one of the" -> "to be in one of the"

mruberry replied:
Generally these errors mention multiple differentiable graphs, but this statement is written as if there's only one.

mruberry commented:
These error messages are much better than what's currently there and the rest of the code looks good, too.

I made a recommendation that might further improve the clarity of the error message, although I may be misunderstanding the cases. Looking forward to hearing your thoughts on it.

Commit:
Update on "Refactored assertAutodiff test to have better error message"


Refactored assertAutodiff to tell exactly which nodes are causing the test to fail with suggestions on how to fix the failure

Differential Revision: [D25212435](https://our.internmc.facebook.com/intern/diff/D25212435)

[ghstack-poisoned]
